pkg/ipcache/types.go

```diff
@@ -76,6 +92,28 @@ func (s prefixInfo) isValid() bool {
 
 func (s prefixInfo) ToLabels() labels.Labels {
 	l := labels.NewLabelsFromModel(nil)
+	// TODO: Actually we should probably just return an error here,
+	//       Then force the caller to pick the identity directly
+	//       This way, the identity reference counting can be entirely
+	//       external to the ipcache metadata map. This helps callers
+	//       understand when to remove the entry from the ipcache map,
+	//       otherwise you either:
+	//       - Add an extra identity allocation for the direct identity,
+	//         meaning that callers cannot decide when they are the "last"
+	//         user of the identity, or
+	//       - Trigger the regular "release the old identity" logic in the
+	//         main TriggerLabelInjection() loop, thereby dropping the
+	//         reference count below zero (which is probably bad?)
+	//       ... The worst danger is some combination where the main loop
+	//       resolves the labels to that identity via this logic, thereby
+	//       adding a reference which can never be decremented because the
+	//       callers remove from the metadata map based on identity refs
+	for _, v := range s {
+		if v.identity != nil {
+			l.MergeLabels(v.labels)
```

Comment:
While in my PR I decided to just ignore the labels if an explicit identity is present, I do wonder also what the expected semantics should be if a prefix both has a bunch of labels and an explicit identity which doesn't necessarily match the identity. My PR falls back to the labelset once the explicit identity is removed, but I'm not if that is actually useful.

Reply:
If the same prefix has an explicit identity and a bunch of labels, the labels should be discarded and the explicit identity take precedence. When removing the explicit identity, the labelset must be fallen back to like your code was doing. Given any particular inputs, it should be possible to calculate exactly what the corresponding outputs (ie ipcache entries) look like, so the fact that an identity override was added then removed should not have an impact on the final result. I think your unit tests in PR https://github.com/cilium/cilium/pull/21667 did a great job of demonstrating this.